src/net/tcp/tcp.cpp

```diff
@@ -509,15 +492,24 @@ bool TCP::unbind(const Socket& socket)
   return false;
 }
 
-void TCP::add_connection(tcp::Connection_ptr conn) {
+bool TCP::add_connection(tcp::Connection_ptr conn) {
   // Stat increment number of incoming connections
   (*incoming_connections_)++;
 
   debug("<TCP::add_connection> Connection added %s \n", conn->to_string().c_str());
-  conn->_on_cleanup({this, &TCP::close_connection});
   conn->bufalloc = mempool_.get_resource();
+
+  // Reject connection if we can't allocate memory
+  if (conn->bufalloc == nullptr
```

Comment:
Maybe `UNLIKELY` here for extra cuteness (could probably be sprinkled on some more places where we check mem).